Function with union type ISE's in SDL

```
type Foo {
  required name: str;
}
type Bar {
  name: str;
}

function name(x: Foo | Bar) -> optional str using (x.name);
```

It works in DDL (https://github.com/geldata/gel/pull/7759).

```
    /home/sully/src/e/edgedb/edb/server/compiler_pool/worker.py, line 169, in compile
        > units, cstate = COMPILER.compile_serialized_request(
    /home/sully/src/e/edgedb/edb/server/compiler/compiler.py, line 607, in compile_serialized_request
        > return self.compile(
    /home/sully/src/e/edgedb/edb/server/compiler/compiler.py, line 686, in compile
        > unit_group = compile(ctx=ctx, source=request.source)
    /home/sully/src/e/edgedb/edb/server/compiler/compiler.py, line 2656, in compile
        > return _try_compile(ctx=ctx, source=source)
    /home/sully/src/e/edgedb/edb/server/compiler/compiler.py, line 2830, in _try_compile
        > return _try_compile_ast(statements=statements, source=source, ctx=ctx)
    /home/sully/src/e/edgedb/edb/server/compiler/compiler.py, line 2882, in _try_compile_ast
        > comp, capabilities = _compile_dispatch_ql(
    /home/sully/src/e/edgedb/edb/server/compiler/compiler.py, line 2561, in _compile_dispatch_ql
        > query = ddl.compile_dispatch_ql_migration(
    /home/sully/src/e/edgedb/edb/server/compiler/ddl.py, line 450, in compile_dispatch_ql_migration
        > return _start_migration(ctx, ql, in_script)
    /home/sully/src/e/edgedb/edb/server/compiler/ddl.py, line 530, in _start_migration
        > target_schema, warnings = s_ddl.apply_sdl(
    /home/sully/src/e/edgedb/edb/schema/ddl.py, line 574, in apply_sdl
        > ddl_stmts = s_decl.sdl_to_ddl(target_schema, documents)
    /home/sully/src/e/edgedb/edb/edgeql/declarative.py, line 448, in sdl_to_ddl
        > trace_dependencies(decl_ast, ctx=tracectx)
    /usr/local/lib/python3.12/functools.py, line 909, in wrapper
        > return dispatch(args[0].__class__)(*args, **kw)
    /home/sully/src/e/edgedb/edb/edgeql/declarative.py, line 1094, in trace_Function
        > assert isinstance(param.type, qlast.TypeName)

AssertionError: 
```

--- a/edb/edgeql/ast.py
+++ b/edb/edgeql/ast.py
@@ -352,11 +352,16 @@ class TypeName(TypeExpr):
     dimensions: typing.Optional[list[int]] = None
 
 
+class TypeOpName(s_enum.StrEnum):
+    OR = '|'
+    AND = '&'
+
+
 class TypeOp(TypeExpr):
     __rust_box__ = {'left', 'right'}
 
     left: TypeExpr
-    op: str
+    op: TypeOpName
     right: TypeExpr
 
 

--- a/edb/edgeql/compiler/schemactx.py
+++ b/edb/edgeql/compiler/schemactx.py
@@ -444,7 +444,7 @@ def get_intersection_type[TypeT: s_types.Type](
 
     targets: Sequence[s_types.Type]
     targets = s_utils.simplify_intersection_types(ctx.env.schema, types)
-    ctx.env.schema, intersection = s_utils.ensure_intersection_type(
+    ctx.env.schema, intersection, _ = s_utils.ensure_intersection_type(
         ctx.env.schema, targets, transient=True
     )
 

--- a/edb/edgeql/compiler/typegen.py
+++ b/edb/edgeql/compiler/typegen.py
@@ -216,13 +216,13 @@ def _ql_typeexpr_get_types(
             # "typeof" and is a list of object types anyway.
             if left_leaf and not left_types[0].is_object_type():
                 raise errors.UnsupportedFeatureError(
-                    f'cannot use type operator {ql_t.op!r} with non-object '
-                    f'type {left_types[0].get_displayname(ctx.env.schema)}',
+                    f"cannot use type operator '{ql_t.op}' with non-object "
+                    f"type {left_types[0].get_displayname(ctx.env.schema)}",
                     span=ql_t.left.span)
             if right_leaf and not right_types[0].is_object_type():
                 raise errors.UnsupportedFeatureError(
-                    f'cannot use type operator {ql_t.op!r} with non-object '
-                    f'type {right_types[0].get_displayname(ctx.env.schema)}',
+                    f"cannot use type operator '{ql_t.op}' with non-object "
+                    f"type {right_types[0].get_displayname(ctx.env.schema)}",
                     span=ql_t.right.span)
 
             # if an operand is either a single type or uses the same operator,

--- a/edb/edgeql/compiler/typegen.py
+++ b/edb/edgeql/compiler/typegen.py
@@ -200,7 +200,7 @@ def _ql_typeexpr_get_types(
         return (None, True, [stype])
 
     elif isinstance(ql_t, qlast.TypeOp):
-        if ql_t.op in ['|', '&']:
+        if ql_t.op in [qlast.TypeOpName.OR, qlast.TypeOpName.AND]:
             (left_op, left_leaf, left_types) = (
                 _ql_typeexpr_get_types(ql_t.left, ctx=ctx)
             )

--- a/edb/edgeql/declarative.py
+++ b/edb/edgeql/declarative.py
@@ -1515,8 +1515,14 @@ def _resolve_type_expr(
 
     elif isinstance(texpr, qlast.TypeOp):
 
-        if texpr.op == '|':
+        if texpr.op == qlast.TypeOpName.OR:
             return qltracer.UnionType([
+                _resolve_type_expr(texpr.left, ctx=ctx),
+                _resolve_type_expr(texpr.right, ctx=ctx),
+            ])
+
+        if texpr.op == qlast.TypeOpName.AND:
+            return qltracer.IntersectionType([
                 _resolve_type_expr(texpr.left, ctx=ctx),
                 _resolve_type_expr(texpr.right, ctx=ctx),
             ])

--- a/edb/edgeql/parser/grammar/expressions.py
+++ b/edb/edgeql/parser/grammar/expressions.py
@@ -2264,12 +2264,18 @@ class NontrivialTypeExpr(Nonterm):
         pass
 
     def reduce_TypeExpr_PIPE_TypeExpr(self, *kids):
-        self.val = qlast.TypeOp(left=kids[0].val, op='|',
-                                right=kids[2].val)
+        self.val = qlast.TypeOp(
+            left=kids[0].val,
+            op=qlast.TypeOpName.OR,
+            right=kids[2].val,
+        )
 
     def reduce_TypeExpr_AMPER_TypeExpr(self, *kids):
-        self.val = qlast.TypeOp(left=kids[0].val, op='&',
-                                right=kids[2].val)
+        self.val = qlast.TypeOp(
+            left=kids[0].val,
+            op=qlast.TypeOpName.AND,
+            right=kids[2].val,
+        )
 
 
 # This is a type expression without angle brackets, so it
@@ -2308,12 +2314,18 @@ class FullTypeExpr(Nonterm):
         pass
 
     def reduce_FullTypeExpr_PIPE_FullTypeExpr(self, *kids):
-        self.val = qlast.TypeOp(left=kids[0].val, op='|',
-                                right=kids[2].val)
+        self.val = qlast.TypeOp(
+            left=kids[0].val,
+            op=qlast.TypeOpName.OR,
+            right=kids[2].val,
+        )
 
     def reduce_FullTypeExpr_AMPER_FullTypeExpr(self, *kids):
-        self.val = qlast.TypeOp(left=kids[0].val, op='&',
-                                right=kids[2].val)
+        self.val = qlast.TypeOp(
+            left=kids[0].val,
+            op=qlast.TypeOpName.AND,
+            right=kids[2].val,
+        )
 
 
 class Subtype(Nonterm):

--- a/edb/edgeql/tracer.py
+++ b/edb/edgeql/tracer.py
@@ -191,14 +191,45 @@ class Alias(ObjectType):
     pass
 
 
-class UnionType(Type):
+class CompositeType(Type):
 
-    def __init__(self, types: list[Type | UnionType | so.Object]) -> None:
+    types: list[Type | CompositeType | s_types.Type]
+
+    def __init__(
+        self,
+        types: list[Type | CompositeType | s_types.Type],
+    ) -> None:
         self.types = types
+
+
+class UnionType(CompositeType):
+
+    def __init__(
+        self,
+        types: list[Type | CompositeType | s_types.Type],
+    ) -> None:
+        super().__init__(types)
 
     def get_name(self, schema: s_schema.Schema) -> sn.QualName:
         component_ids = sorted(str(t.get_name(schema)) for t in self.types)
         nqname = f"({' | '.join(component_ids)})"
+        return sn.QualName(name=nqname, module='__derived__')
+
+    def is_object_type(self) -> bool:
+        return True
+
+
+class IntersectionType(CompositeType):
+
+    def __init__(
+        self,
+        types: list[Type | CompositeType | s_types.Type],
+    ) -> None:
+        super().__init__(types)
+
+    def get_name(self, schema: s_schema.Schema) -> sn.QualName:
+        component_ids = sorted(str(t.get_name(schema)) for t in self.types)
+        nqname = f"({' & '.join(component_ids)})"
         return sn.QualName(name=nqname, module='__derived__')
 
     def is_object_type(self) -> bool:
@@ -1035,15 +1066,25 @@ def _resolve_type_expr(
 
     elif isinstance(texpr, qlast.TypeOp):
 
-        if texpr.op == '|':
-            return UnionType([
-                _resolve_type_expr(texpr.left, ctx=ctx),
-                _resolve_type_expr(texpr.right, ctx=ctx),
-            ])
+        left = _resolve_type_expr(texpr.left, ctx=ctx)
+        right = _resolve_type_expr(texpr.right, ctx=ctx)
 
+        ThisCompositeType: type[CompositeType] = (
+            UnionType
+            if texpr.op == qlast.TypeOpName.OR else
+            IntersectionType
+        )
+
+        if isinstance(left, ThisCompositeType):
+            if isinstance(right, ThisCompositeType):
+                return ThisCompositeType(left.types + right.types)
+            else:
+                return ThisCompositeType(left.types + [right])
         else:
-            raise NotImplementedError(
-                f'unsupported type operation: {texpr.op}')
+            if isinstance(right, ThisCompositeType):
+                return ThisCompositeType([left] + right.types)
+            else:
+                return ThisCompositeType([left, right])
 
     else:
         raise NotImplementedError(

--- a/edb/pgsql/delta.py
+++ b/edb/pgsql/delta.py
@@ -3700,6 +3700,14 @@ class CreateUnionType(
     pass
 
 
+class CreateIntersectionType(
+    MetaCommand,
+    adapts=s_types.CreateIntersectionType,
+    metaclass=CommandMeta,
+):
+    pass
+
+
 class ObjectTypeMetaCommand(AliasCapableMetaCommand, CompositeMetaCommand):
     def schedule_endpoint_delete_action_update(self, obj, schema, context):
         endpoint_delete_actions = context.get(

--- a/edb/pgsql/delta.py
+++ b/edb/pgsql/delta.py
@@ -2741,8 +2741,11 @@ class AlterScalarType(ScalarTypeMetaCommand, adapts=s_scalars.AlterScalarType):
         for prop, new_typ in props.items():
             try:
                 cmd.add(new_typ.as_create_delta(schema))
-            except errors.UnsupportedFeatureError:
-                pass
+            except NotImplementedError as e:
+                if e.args == ('unsupported typeshell',):
+                    pass
+                else:
+                    raise
 
             if prop.get_default(schema):
                 delta_alter, cmd_alter, _alter_context = prop.init_delta_branch(

--- a/edb/schema/objtypes.py
+++ b/edb/schema/objtypes.py
@@ -412,7 +412,7 @@ def get_or_create_intersection_type(
     *,
     module: Optional[str] = None,
     transient: bool = False,
-) -> tuple[s_schema.Schema, ObjectType]:
+) -> tuple[s_schema.Schema, ObjectType, bool]:
 
     name = s_types.get_intersection_type_name(
         (c.get_name(schema) for c in components),
@@ -420,6 +420,7 @@ def get_or_create_intersection_type(
     )
 
     objtype = schema.get(name, default=None, type=ObjectType)
+    created = objtype is None
     if objtype is None:
         components = list(components)
 
@@ -463,7 +464,7 @@ def get_or_create_intersection_type(
                 schema = objtype.add_pointer(schema, ptr)
 
     assert isinstance(objtype, ObjectType)
-    return schema, objtype
+    return schema, objtype, created
 
 
 class ObjectTypeCommandContext(

--- a/edb/schema/pointers.py
+++ b/edb/schema/pointers.py
@@ -1219,6 +1219,19 @@ class PointerCommandOrFragment[Pointer_T: Pointer](
             self.set_attribute_value('linkful', True)
 
         if inf_target_ref is not None:
+            if inf_target_ref.has_intersection():
+                raise errors.UnsupportedFeatureError(
+                    (
+                        f'unsupported type intersection in schema '
+                        f'{inf_target_ref.get_name(schema).name}'
+                    ),
+                    hint=(
+                        f'Type intersections are currently '
+                        f'unsupported as valid link targets.'
+                    ),
+                    span=self.span,
+                )
+
             span = self.get_attribute_span('target')
             self.set_attribute_value(
                 'target',

--- a/edb/schema/pointers.py
+++ b/edb/schema/pointers.py
@@ -3179,7 +3179,7 @@ def get_or_create_union_pointer(
             cardinality = qltypes.SchemaCardinality.Many
             break
 
-    required = all(component.get_required(schema) for component in components)
+    required = any(component.get_required(schema) for component in components)
     metacls = type(components[0])
     default_base_name = metacls.get_default_base_name()
     assert default_base_name is not None
@@ -3245,10 +3245,11 @@ def get_or_create_intersection_pointer(
     if len(components) == 1:
         return schema, components[0]
 
+    required = all(component.get_required(schema) for component in components)
     targets: Sequence[s_types.Type]
     targets = list(filter(None, [p.get_target(schema) for p in components]))
     targets = utils.simplify_intersection_types(schema, targets)
-    schema, target = utils.ensure_intersection_type(
+    schema, target, _ = utils.ensure_intersection_type(
         schema, targets, module=modname)
 
     cardinality = qltypes.SchemaCardinality.One
@@ -3270,6 +3271,7 @@ def get_or_create_intersection_pointer(
         attrs={
             'intersection_of': so.ObjectSet.create(schema, components),
             'cardinality': cardinality,
+            'required': required,
         },
         transient=transient,
     )

--- a/edb/schema/pointers.py
+++ b/edb/schema/pointers.py
@@ -3192,7 +3192,7 @@ def get_or_create_union_pointer(
             cardinality = qltypes.SchemaCardinality.Many
             break
 
-    required = any(component.get_required(schema) for component in components)
+    required = all(component.get_required(schema) for component in components)
     metacls = type(components[0])
     default_base_name = metacls.get_default_base_name()
     assert default_base_name is not None
@@ -3258,17 +3258,17 @@ def get_or_create_intersection_pointer(
     if len(components) == 1:
         return schema, components[0]
 
-    required = all(component.get_required(schema) for component in components)
+    required = any(component.get_required(schema) for component in components)
     targets: Sequence[s_types.Type]
     targets = list(filter(None, [p.get_target(schema) for p in components]))
     targets = utils.simplify_intersection_types(schema, targets)
     schema, target, _ = utils.ensure_intersection_type(
         schema, targets, module=modname)
 
-    cardinality = qltypes.SchemaCardinality.One
+    cardinality = qltypes.SchemaCardinality.Many
     for component in components:
-        if component.get_cardinality(schema) is qltypes.SchemaCardinality.Many:
-            cardinality = qltypes.SchemaCardinality.Many
+        if component.get_cardinality(schema) is qltypes.SchemaCardinality.One:
+            cardinality = qltypes.SchemaCardinality.One
             break
 
     metacls = type(components[0])

--- a/edb/schema/types.py
+++ b/edb/schema/types.py
@@ -636,12 +636,10 @@ class TypeShell(so.ObjectShell[TypeT_co]):
         view_name: Optional[s_name.QualName] = None,
         attrs: Optional[dict[str, Any]] = None,
     ) -> sd.Command:
-        raise errors.UnsupportedFeatureError(
-            f'unsupported type intersection in schema {str(view_name)}',
-            hint=f'Type intersections are currently '
-                 f'unsupported as valid link targets.',
-            span=self.span,
-        )
+        raise NotImplementedError('unsupported typeshell')
+
+    def has_intersection(self) -> bool:
+        return False
 
 
 class TypeExprShell(TypeShell[TypeT_co]):
@@ -674,6 +672,12 @@ class TypeExprShell(TypeShell[TypeT_co]):
         schema: s_schema.Schema,
     ) -> tuple[TypeShell[TypeT_co], ...]:
         return self.components
+
+    def has_intersection(self) -> bool:
+        return any(
+            c.has_intersection()
+            for c in self.components
+        )
 
 
 class UnionTypeShell(TypeExprShell[TypeT_co]):
@@ -1038,6 +1042,9 @@ class IntersectionTypeShell(TypeExprShell[TypeT_co]):
         cmd.set_attribute_value('components', tuple(self.components))
         cmd.set_attribute_value('span', self.span)
         return cmd
+
+    def has_intersection(self) -> bool:
+        return True
 
 
 _collection_impls: dict[str, type[Collection]] = {}

--- a/edb/schema/types.py
+++ b/edb/schema/types.py
@@ -709,6 +709,9 @@ class UnionTypeShell(TypeExprShell[TypeT_co]):
     ) -> sd.Command:
         assert isinstance(self.name, s_name.QualName)
         cmd = CreateUnionType(classname=self.name)
+        for component in self.components:
+            if isinstance(component, TypeExprShell):
+                cmd.add_prerequisite(component.as_create_delta(schema))
         cmd.set_attribute_value('name', self.name)
         cmd.set_attribute_value('components', tuple(self.components))
         cmd.set_attribute_value('is_opaque_union', self.opaque)
@@ -894,6 +897,9 @@ class CreateUnionType(sd.CreateObject[InheritingType], CompoundTypeCommand):
 
         from edb.schema import types as s_types
 
+        for cmd in self.get_prerequisites():
+            schema = cmd.apply(schema, context)
+
         if not context.canonical:
             components: Sequence[s_types.Type] = [
                 c.resolve(schema)
@@ -929,7 +935,67 @@ class CreateUnionType(sd.CreateObject[InheritingType], CompoundTypeCommand):
 
                 self.add(delta)
 
-        for cmd in self.get_subcommands():
+        for cmd in self.get_subcommands(include_prerequisites=False):
+            schema = cmd.apply(schema, context)
+
+        return schema
+
+
+class CreateIntersectionType(
+    sd.CreateObject[InheritingType], CompoundTypeCommand
+):
+
+    def apply(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+    ) -> s_schema.Schema:
+
+        from edb.schema import types as s_types
+
+        for cmd in self.get_prerequisites():
+            schema = cmd.apply(schema, context)
+
+        if not context.canonical:
+            components: Sequence[s_types.Type] = [
+                c.resolve(schema)
+                for c in self.get_attribute_value('components')
+            ]
+
+            try:
+                new_schema, intersection_type, created = (
+                    utils.ensure_intersection_type(
+                        schema,
+                        components,
+                        module=self.classname.module,
+                    )
+                )
+            except errors.SchemaError as e:
+                intersection_name = (
+                    '(' + ' | '.join(sorted(
+                    c.get_displayname(schema)
+                    for c in components
+                    )) + ')'
+                )
+                e.args = (
+                    (
+                        f'cannot create intersection '
+                        f'{intersection_name} {e.args[0]}',
+                    )
+                    + e.args[1:]
+                )
+                e.set_span(self.get_attribute_value('span'))
+                raise e
+
+            if created:
+                delta = intersection_type.as_create_delta(
+                    schema=new_schema,
+                    context=so.ComparisonContext(),
+                )
+
+                self.add(delta)
+
+        for cmd in self.get_subcommands(include_prerequisites=False):
             schema = cmd.apply(schema, context)
 
         return schema
@@ -955,6 +1021,23 @@ class IntersectionTypeShell(TypeExprShell[TypeT_co]):
             schemaclass=schemaclass,
             span=span
         )
+
+    def as_create_delta(
+        self,
+        schema: s_schema.Schema,
+        *,
+        view_name: Optional[s_name.QualName] = None,
+        attrs: Optional[dict[str, Any]] = None,
+    ) -> sd.Command:
+        assert isinstance(self.name, s_name.QualName)
+        cmd = CreateIntersectionType(classname=self.name)
+        for component in self.components:
+            if isinstance(component, TypeExprShell):
+                cmd.add_prerequisite(component.as_create_delta(schema))
+        cmd.set_attribute_value('name', self.name)
+        cmd.set_attribute_value('components', tuple(self.components))
+        cmd.set_attribute_value('span', self.span)
+        return cmd
 
 
 _collection_impls: dict[str, type[Collection]] = {}

--- a/edb/schema/utils.py
+++ b/edb/schema/utils.py
@@ -408,7 +408,7 @@ def type_op_ast_to_type_shell[TypeT: s_types.Type](
 
     from . import types as s_types
 
-    if node.op != '|':
+    if node.op not in [qlast.TypeOpName.OR, qlast.TypeOpName.AND]:
         raise errors.UnsupportedFeatureError(
             f'unsupported type expression operator: {node.op}',
             span=node.span,
@@ -438,31 +438,47 @@ def type_op_ast_to_type_shell[TypeT: s_types.Type](
         schema=schema,
     )
 
-    if isinstance(left, s_types.UnionTypeShell):
-        if isinstance(right, s_types.UnionTypeShell):
-            return s_types.UnionTypeShell(
+    CompositeTypeShell = (
+        s_types.UnionTypeShell
+        if node.op == qlast.TypeOpName.OR else
+        s_types.IntersectionTypeShell
+    )
+
+    # Doubled check for s_types.TypeExprShell to reassure mypy
+    if (
+        isinstance(left, CompositeTypeShell)
+        and isinstance(left, s_types.TypeExprShell)
+    ):
+        if (
+            isinstance(right, CompositeTypeShell)
+            and isinstance(right, s_types.TypeExprShell)
+        ):
+            return CompositeTypeShell(
                 components=left.components + right.components,
                 module=module,
                 schemaclass=metaclass,
                 span=node.span,
             )
         else:
-            return s_types.UnionTypeShell(
+            return CompositeTypeShell(
                 components=left.components + (right,),
                 module=module,
                 schemaclass=metaclass,
                 span=node.span,
             )
     else:
-        if isinstance(right, s_types.UnionTypeShell):
-            return s_types.UnionTypeShell(
+        if (
+            isinstance(right, CompositeTypeShell)
+            and isinstance(right, s_types.TypeExprShell)
+        ):
+            return CompositeTypeShell(
                 components=(left,) + right.components,
                 schemaclass=metaclass,
                 module=module,
                 span=node.span,
             )
         else:
-            return s_types.UnionTypeShell(
+            return CompositeTypeShell(
                 components=(left, right),
                 module=module,
                 schemaclass=metaclass,
@@ -567,19 +583,32 @@ def typeref_to_ast(
                 for st in t.get_subtypes(schema)
             ]
         )
-    elif isinstance(t, s_types.Type) and t.is_union_type(schema):
-        object_set = t.get_union_of(schema)
+    elif (
+        isinstance(t, s_types.Type)
+        and (t.is_union_type(schema) or t.is_intersection_type(schema))
+    ):
+        object_set = (
+            t.get_union_of(schema)
+            if t.is_union_type(schema) else
+            t.get_intersection_of(schema)
+        )
         assert object_set is not None
 
         component_objects = tuple(object_set.objects(schema))
-        result = typeref_to_ast(schema, component_objects[0],
-                                disambiguate_std=disambiguate_std)
+        result = typeref_to_ast(
+            schema, component_objects[0], disambiguate_std=disambiguate_std
+        )
         for component_object in component_objects[1:]:
             result = qlast.TypeOp(
                 left=result,
-                op='|',
-                right=typeref_to_ast(schema, component_object,
-                                     disambiguate_std=disambiguate_std),
+                op=(
+                    qlast.TypeOpName.OR
+                    if t.is_union_type(schema) else
+                    qlast.TypeOpName.AND
+                ),
+                right=typeref_to_ast(
+                    schema, component_object, disambiguate_std=disambiguate_std
+                ),
             )
     elif isinstance(t, so.QualifiedObject):
         t_name = t.get_name(schema)
@@ -711,7 +740,16 @@ def type_shell_to_ast(
         for component in components[1:]:
             result = qlast.TypeOp(
                 left=result,
-                op='|',
+                op=qlast.TypeOpName.OR,
+                right=typeref_to_ast(schema, component),
+            )
+    elif isinstance(t, s_types.IntersectionTypeShell):
+        components = t.get_components(schema)
+        result = typeref_to_ast(schema, components[0])
+        for component in components[1:]:
+            result = qlast.TypeOp(
+                left=result,
+                op=qlast.TypeOpName.AND,
                 right=typeref_to_ast(schema, component),
             )
     elif isinstance(t, s_scalars.AnonymousEnumTypeShell):
@@ -1305,12 +1343,12 @@ def ensure_intersection_type(
     *,
     transient: bool = False,
     module: Optional[str] = None,
-) -> tuple[s_schema.Schema, s_types.Type]:
+) -> tuple[s_schema.Schema, s_types.Type, bool]:
 
     from edb.schema import objtypes as s_objtypes
 
     if len(types) == 1:
-        return schema, next(iter(types))
+        return schema, next(iter(types)), False
 
     seen_scalars = False
     seen_objtypes = False

--- a/tests/schemas/cards_ir_inference.esdl
+++ b/tests/schemas/cards_ir_inference.esdl
@@ -165,3 +165,16 @@ type Named2Sub extending Named2;
 global Alice := (select User filter .name = 'Alice');
 
 permission GameAdmin;
+
+type TypeExprA {
+    val: str;
+};
+type TypeExprB {
+    required val: str;
+};
+type TypeExprC {
+    multi val: str;
+};
+type TypeExprD {
+    required multi val: str;
+};

--- a/tests/test_edgeql_ir_card_inference.py
+++ b/tests/test_edgeql_ir_card_inference.py
@@ -1349,3 +1349,103 @@ class TestEdgeQLCardinalityInference(tb.BaseEdgeQLCompilerTest):
 % OK %
         ONE
         """
+
+    def test_edgeql_ir_card_inference_158(self):
+        """
+        with TypeExpr := Object[is TypeExprA | TypeExprB]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        AT_MOST_ONE
+        """
+
+    def test_edgeql_ir_card_inference_159(self):
+        """
+        with TypeExpr := Object[is TypeExprA | TypeExprC]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        MANY
+        """
+
+    def test_edgeql_ir_card_inference_160(self):
+        """
+        with TypeExpr := Object[is TypeExprA | TypeExprD]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        MANY
+        """
+
+    def test_edgeql_ir_card_inference_161(self):
+        """
+        with TypeExpr := Object[is TypeExprB | TypeExprC]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        MANY
+        """
+
+    def test_edgeql_ir_card_inference_162(self):
+        """
+        with TypeExpr := Object[is TypeExprB | TypeExprD]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        AT_LEAST_ONE
+        """
+
+    def test_edgeql_ir_card_inference_163(self):
+        """
+        with TypeExpr := Object[is TypeExprC | TypeExprD]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        MANY
+        """
+
+    def test_edgeql_ir_card_inference_164(self):
+        """
+        with TypeExpr := Object[is TypeExprA & TypeExprB]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        ONE
+        """
+
+    @tb.must_fail(errors.SchemaError, "different cardinality")
+    def test_edgeql_ir_card_inference_165(self):
+        """
+        with TypeExpr := Object[is TypeExprA & TypeExprC]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        AT_MOST_ONE
+        """
+
+    @tb.must_fail(errors.SchemaError, "different cardinality")
+    def test_edgeql_ir_card_inference_166(self):
+        """
+        with TypeExpr := Object[is TypeExprA & TypeExprD]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        ONE
+        """
+
+    @tb.must_fail(errors.SchemaError, "different cardinality")
+    def test_edgeql_ir_card_inference_167(self):
+        """
+        with TypeExpr := Object[is TypeExprB & TypeExprC]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        ONE
+        """
+
+    @tb.must_fail(errors.SchemaError, "different cardinality")
+    def test_edgeql_ir_card_inference_168(self):
+        """
+        with TypeExpr := Object[is TypeExprB & TypeExprD]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        ONE
+        """
+
+    def test_edgeql_ir_card_inference_169(self):
+        """
+        with TypeExpr := Object[is TypeExprC & TypeExprD]
+        select assert_exists(assert_single(TypeExpr)).val
+% OK %
+        AT_LEAST_ONE
+        """

--- a/tests/test_schema.py
+++ b/tests/test_schema.py
@@ -2150,11 +2150,54 @@ class TestSchema(tb.BaseSchemaLoadTest):
         self.assertIsNone(A_tf_d.maybe_get_ptr(schema, 'd_prop2'))
         A_tf_df.getptr(schema, s_name.UnqualName('df_prop'))
 
-    def test_schema_advanced_function_params(self):
+    def test_schema_advanced_function_params_01(self):
         """
-            type Foo { required value: int64 };
-            type Bar { required value: int64 };
-            function get_value(x: Foo | Bar) -> int64 using (x.value);
+            type A { required value: int64 };
+            type B { required value: int64 };
+            function get_value(x: A | B) -> int64 using (x.value);
+        """
+
+    def test_schema_advanced_function_params_02(self):
+        """
+            type A { required value: int64 };
+            type B { required value: int64 };
+            function get_value(x: A & B) -> int64 using (x.value);
+        """
+
+    def test_schema_advanced_function_params_03(self):
+        """
+            type A { required value: int64 };
+            type B { required value: int64 };
+            type C { required value: int64 };
+            type D { required value: int64 };
+            function get_value(x: (A | B) | (C | D)) -> int64 using (x.value);
+        """
+
+    def test_schema_advanced_function_params_04(self):
+        """
+            type A { required value: int64 };
+            type B { required value: int64 };
+            type C { required value: int64 };
+            type D { required value: int64 };
+            function get_value(x: (A & B) & (C & D)) -> int64 using (x.value);
+        """
+
+    def test_schema_advanced_function_params_05(self):
+        """
+            type A { required value: int64 };
+            type B { required value: int64 };
+            type C { required value: int64 };
+            type D { required value: int64 };
+            function get_value(x: (A & B) | (C & D)) -> int64 using (x.value);
+        """
+
+    def test_schema_advanced_function_params_06(self):
+        """
+            type A { required value: int64 };
+            type B { required value: int64 };
+            type C { required value: int64 };
+            type D { required value: int64 };
+            function get_value(x: (A | B) & (C | D)) -> int64 using (x.value);
         """
 
     def test_schema_ancestor_propagation_on_sdl_migration(self):

